src/stirling/source_connectors/socket_tracer/protocols/amqp/parse.cc

```diff
@@ -90,7 +90,7 @@ ParseState ParseFrame(message_type_t type, std::string_view* buf, Frame* packet)
   packet->frame_type = frame_type;
   packet->channel = channel;
   packet->payload_size = payload_size;
-  if (decoder.BufSize() < payload_size) {
+  if (decoder.BufSize() <= payload_size) {
     return ParseState::kInvalid;
   }
```

Comment:
Have you considered if `ParseState::kInvalid` is correct here? If the buffer doesn't contain enough data, then shouldn't this condition be `ParseState::kNeedsMoreData` instead?

Reply:
I agree (and had not considered that).

Reply:
Updated PR & also verbiage in the PR summary.

Reply:
If it's not too much trouble, it would be great to see the stirling wrapper output with `return ParseState::kInvalid;` against ` vs ParseState::kNeedsMoreData`. I would expect that the former would have less amqp frames than the latter and it would give us additional verification that the return value is correct (since we don't have test coverage yet).

Reply:
We edited `stirling_wrapper.cc` to print some stats about how many amqp events we parsed. Comparing results between `kInvalid` and `kNeedsMoreData` are currently inconclusive. 

Here is a snippet of the log file from the experiment running with `kNeedsMoreData`:
```I20230303 00:59:37.209316 209328 stirling_wrapper.cc:100] batch_idx: 58, total num. amqp_events: 185767, num. new amqp_events: 1374
I20230303 00:59:37.745893 209328 stirling_wrapper.cc:100] batch_idx: 59, total num. amqp_events: 187141, num. new amqp_events: 1239
I20230303 00:59:38.043551 209328 stirling_wrapper.cc:100] batch_idx: 60, total num. amqp_events: 188380, num. new amqp_events: 1398
I20230303 00:59:38.340513 209328 stirling_wrapper.cc:100] batch_idx: 61, total num. amqp_events: 189778, num. new amqp_events: 1257
I20230303 00:59:38.640245 209328 stirling_wrapper.cc:100] batch_idx: 62, total num. amqp_events: 191035, num. new amqp_events: 1335
I20230303 00:59:38.908176 209328 stirling_wrapper.cc:100] batch_idx: 63, total num. amqp_events: 192370, num. new amqp_events: 1650
I20230303 00:59:56.498466 209328 stirling_wrapper.cc:100] batch_idx: 64, total num. amqp_events: 194020, num. new amqp_events: 936
I20230303 00:59:59.804971 209328 stirling_wrapper.cc:100] batch_idx: 65, total num. amqp_events: 194956, num. new amqp_events: 1240
I20230303 01:00:00.808861 209328 stirling_wrapper.cc:100] batch_idx: 66, total num. amqp_events: 196196, num. new amqp_events: 78
I20230303 01:00:01.835283 209328 stirling_wrapper.cc:100] batch_idx: 67, total num. amqp_events: 196274, num. new amqp_events: 60
I20230303 01:00:02.843797 209328 stirling_wrapper.cc:100] batch_idx: 68, total num. amqp_events: 196334, num. new amqp_events: 72
I20230303 01:00:03.934898 209328 stirling_wrapper.cc:100] batch_idx: 69, total num. amqp_events: 196406, num. new amqp_events: 84
I20230303 01:00:05.059293 209328 stirling_wrapper.cc:100] batch_idx: 70, total num. amqp_events: 196490, num. new amqp_events: 126
I20230303 01:00:06.151943 209328 stirling_wrapper.cc:100] batch_idx: 71, total num. amqp_events: 196616, num. new amqp_events: 66
I20230303 01:00:07.187829 209328 stirling_wrapper.cc:100] batch_idx: 72, total num. amqp_events: 196682, num. new amqp_events: 24
I20230303 01:00:08.213675 209328 stirling_wrapper.cc:100] batch_idx: 73, total num. amqp_events: 196706, num. new amqp_events: 18
I20230303 01:00:10.337916 209328 stirling_wrapper.cc:100] batch_idx: 74, total num. amqp_events: 196724, num. new amqp_events: 156
I20230303 01:00:11.342752 209328 stirling_wrapper.cc:100] batch_idx: 75, total num. amqp_events: 196880, num. new amqp_events: 156
I20230303 01:00:12.345613 209328 stirling_wrapper.cc:100] batch_idx: 76, total num. amqp_events: 197036, num. new amqp_events: 54
I20230303 01:00:13.354014 209328 stirling_wrapper.cc:100] batch_idx: 77, total num. amqp_events: 197090, num. new amqp_events: 84
I20230303 01:00:14.374197 209328 stirling_wrapper.cc:100] batch_idx: 78, total num. amqp_events: 197174, num. new amqp_events: 30
```

It seems that, upon startup, we parse about 1000 to 2000 amqp events per batch, but later that number drops to 50 to 150 per batch. But 50~100 is the same amount as parsed when using `kInvalid`. Without knowing the actual expected event rate for `amqp`, it is hard to draw a conclusion at this time.

Reply:
Pete and I discussed offline that even though the test above was inconclusive, we believe `kNeedsMoreData` is correct. Once we have a chance to add a test case (https://github.com/pixie-io/pixie/issues/945), we will be able to verify that but in the short term we need to prevent the seg fault from happening.